File: examples/consumers/go-gin/internal/webhouse/webhouse.go
// Package webhouse is the F125 reference reader for @webhouse/cms file-based content.
//
// It reads JSON documents from content/{collection}/ and exposes them as
// Document instances. Designed to be thin (stdlib + zero deps), and safe —
// slugs and collection names are validated to prevent path traversal.
//
// Reference implementation for the future github.com/webhousecode/cms-reader-go
// module.
package webhouse

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"sync"
)

// Document mirrors a single @webhouse/cms JSON file.
type Document struct {
	ID               string                 `json:"id"`
	Slug             string                 `json:"slug"`
	Status           string                 `json:"status"`
	Locale           string                 `json:"locale,omitempty"`
	TranslationGroup string                 `json:"translationGroup,omitempty"`
	Data             map[string]interface{} `json:"data"`
	CreatedAt        string                 `json:"createdAt,omitempty"`
	UpdatedAt        string                 `json:"updatedAt,omitempty"`
}

// IsPublished returns true when the document status is "published".
func (d *Document) IsPublished() bool { return d.Status == "published" }

// String safely extracts a string field from data.
func (d *Document) String(key string) string {
	if d.Data == nil {
		return ""
	}
	if v, ok := d.Data[key].(string); ok {
		return v
	}
	return ""
}

// StringOr returns the string field or a fallback default.
func (d *Document) StringOr(key, fallback string) string {
	if v := d.String(key); v != "" {
		return v
	}
	return fallback
}

// StringSlice extracts an array of strings (e.g. tags).
func (d *Document) StringSlice(key string) []string {
	if d.Data == nil {
		return nil
	}
	raw, ok := d.Data[key].([]interface{})
	if !ok {
		return nil
	}
	out := make([]string, 0, len(raw))
	for _, v := range raw {
		if s, ok := v.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

// Reader provides typed access to a content/ directory.
type Reader struct {
	contentDir string

	mu    sync.RWMutex
	cache map[string][]Document // collection:locale -> documents
}

// safeName matches the same pattern enforced by CMS admin and other readers.
var safeName = regexp.MustCompile(`^[a-z0-9]([a-z0-9-]*[a-z0-9])?$`)

// ErrInvalidName is returned when a collection name or slug fails validation.
var ErrInvalidName = errors.New("webhouse: invalid name (must match ^[a-z0-9][a-z0-9-]*$)")

// New constructs a Reader pointed at the given content directory.
func New(contentDir string) *Reader {
	abs, _ := filepath.Abs(contentDir)
	return &Reader{
		contentDir: abs,
		cache:      make(map[string][]Document),
	}
}

func validate(name string) error {
	if !safeName.MatchString(name) {
		return ErrInvalidName
	}
	return nil
}

// Collection lists all published documents in a collection. If locale is
// empty, all locales are returned. Sorted by data.date descending.
func (r *Reader) Collection(collection, locale string) ([]Document, error) {
	if err := validate(collection); err != nil {
		return nil, err
	}

	dir := filepath.Join(r.contentDir, collection)
	entries, err := os.ReadDir(dir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}

	var docs []Document
	for _, e := range entries {
		if !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		raw, err := os.ReadFile(filepath.Join(dir, e.Name()))
		if err != nil {
			continue
		}
		var d Document
		if err := json.Unmarshal(raw, &d); err != nil {
			continue
		}
		if !d.IsPublished() {
			continue
		}
		if locale != "" && d.Locale != locale {
			continue
		}
		docs = append(docs, d)
	}

	sort.SliceStable(docs, func(i, j int) bool {
		return docs[i].StringOr("date", "") > docs[j].StringOr("date", "")
	})
	return docs, nil
}

// Document loads a single published document by collection and slug.
func (r *Reader) Document(collection, slug string) (*Document, error) {
	if err := validate(collection); err != nil {
		return nil, err
	}
	if err := validate(slug); err != nil {
		return nil, err
	}

	path := filepath.Join(r.contentDir, collection, slug+".json")
	abs, err := filepath.Abs(path)
	if err != nil {
		return nil, err
	}
	if !strings.HasPrefix(abs, r.contentDir) {
		return nil, ErrInvalidName
	}

	raw, err := os.ReadFile(abs)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var d Document
	if err := json.Unmarshal(raw, &d); err != nil {
		return nil, err
	}
	if !d.IsPublished() {
		return nil, nil
	}
	return &d, nil
}

// FindTranslation resolves the sibling translation of a document via its
// translationGroup. Returns nil if there is none in another locale.
func (r *Reader) FindTranslation(doc *Document, collection string) (*Document, error) {
	if doc == nil || doc.TranslationGroup == "" {
		return nil, nil
	}
	all, err := r.Collection(collection, "")
	if err != nil {
		return nil, err
	}
	for i := range all {
		o := &all[i]
		if o.TranslationGroup == doc.TranslationGroup && o.Locale != doc.Locale {
			return o, nil
		}
	}
	return nil, nil
}

File: examples/consumers/rust-axum-blog/src/webhouse.rs
//! F125 reference reader for @webhouse/cms file-based content (Rust).
//!
//! Reads JSON documents from content/{collection}/ via std::fs.
//! Designed to be thin (zero deps beyond serde/serde_json) and safe —
//! slugs and collection names are validated to prevent path traversal.

use once_cell::sync::Lazy;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::RwLock;

static SAFE_NAME: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^[a-z0-9]([a-z0-9-]*[a-z0-9])?$").unwrap()
});

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Document {
    #[serde(default)]
    pub id: String,
    pub slug: String,
    pub status: String,
    #[serde(default)]
    pub locale: Option<String>,
    #[serde(default, rename = "translationGroup")]
    pub translation_group: Option<String>,
    #[serde(default)]
    pub data: HashMap<String, serde_json::Value>,
    #[serde(default, rename = "createdAt")]
    pub created_at: Option<String>,
    #[serde(default, rename = "updatedAt")]
    pub updated_at: Option<String>,
}

impl Document {
    pub fn is_published(&self) -> bool {
        self.status == "published"
    }

    pub fn string(&self, key: &str) -> String {
        self.data.get(key)
            .and_then(|v| v.as_str())
            .map(|s| s.to_string())
            .unwrap_or_default()
    }

    pub fn string_or(&self, key: &str, default: &str) -> String {
        let v = self.string(key);
        if v.is_empty() { default.to_string() } else { v }
    }
}

#[derive(Debug)]
pub enum WebhouseError {
    InvalidName(String),
    Io(std::io::Error),
    Parse(serde_json::Error),
}

impl std::fmt::Display for WebhouseError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::InvalidName(n) => write!(f, "invalid name '{}'", n),
            Self::Io(e) => write!(f, "io error: {}", e),
            Self::Parse(e) => write!(f, "json parse error: {}", e),
        }
    }
}

impl std::error::Error for WebhouseError {}

pub struct Reader {
    content_dir: PathBuf,
    globals_cache: RwLock<Option<Document>>,
}

impl Reader {
    pub fn new<P: AsRef<Path>>(content_dir: P) -> Self {
        Self {
            content_dir: content_dir.as_ref().canonicalize()
                .unwrap_or_else(|_| content_dir.as_ref().to_path_buf()),
            globals_cache: RwLock::new(None),
        }
    }

    pub fn collection(&self, name: &str, locale: Option<&str>) -> Result<Vec<Document>, WebhouseError> {
        validate(name)?;
        let dir = self.content_dir.join(name);
        if !dir.is_dir() {
            return Ok(Vec::new());
        }

        let mut docs = Vec::new();
        for entry in fs::read_dir(&dir).map_err(WebhouseError::Io)? {
            let entry = entry.map_err(WebhouseError::Io)?;
            let path = entry.path();
            if path.extension().and_then(|s| s.to_str()) != Some("json") {
                continue;
            }
            let raw = match fs::read_to_string(&path) {
                Ok(s) => s,
                Err(_) => continue,
            };
            let doc: Document = match serde_json::from_str(&raw) {
                Ok(d) => d,
                Err(_) => continue,
            };
            if !doc.is_published() {
                continue;
            }
            if let Some(loc) = locale {
                if doc.locale.as_deref() != Some(loc) {
                    continue;
                }
            }
            docs.push(doc);
        }

        docs.sort_by(|a, b| b.string_or("date", "").cmp(&a.string_or("date", "")));
        Ok(docs)
    }

    pub fn document(&self, collection: &str, slug: &str) -> Result<Option<Document>, WebhouseError> {
        validate(collection)?;
        validate(slug)?;

        let path = self.content_dir.join(collection).join(format!("{}.json", slug));
        let canonical = match path.canonicalize() {
            Ok(p) => p,
            Err(_) => return Ok(None),
        };
        if !canonical.starts_with(&self.content_dir) {
            return Err(WebhouseError::InvalidName(slug.to_string()));
        }

        let raw = fs::read_to_string(&canonical).map_err(WebhouseError::Io)?;
        let doc: Document = serde_json::from_str(&raw).map_err(WebhouseError::Parse)?;
        if !doc.is_published() {
            return Ok(None);
        }
        Ok(Some(doc))
    }

    pub fn find_translation(&self, doc: &Document, collection: &str) -> Result<Option<Document>, WebhouseError> {
        let tg = match &doc.translation_group {
            Some(t) => t,
            None => return Ok(None),
        };
        let all = self.collection(collection, None)?;
        Ok(all.into_iter().find(|other| {
            other.translation_group.as_deref() == Some(tg.as_str())
                && other.locale != doc.locale
        }))
    }

    pub fn globals(&self) -> Document {
        {
            let cache = self.globals_cache.read().unwrap();
            if let Some(g) = cache.as_ref() {
                return g.clone();
            }
        }
        let g = self.document("globals", "site").ok().flatten().unwrap_or_else(|| Document {
            id: String::new(),
            slug: "site".to_string(),
            status: "published".to_string(),
            locale: None,
            translation_group: None,
            data: HashMap::new(),
            created_at: None,
            updated_at: None,
        });
        *self.globals_cache.write().unwrap() = Some(g.clone());
        g
    }
}

fn validate(name: &str) -> Result<(), WebhouseError> {
    if !SAFE_NAME.is_match(name) {
        return Err(WebhouseError::InvalidName(name.to_string()));
    }
    Ok(())
}
